File: LinkedList/C/E206-reverse-linked-list.c
// time:  O(n)
// space: O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *	   int val;
 *	   struct ListNode *next;
 * };
 */
// Iteratively
struct ListNode* reverseList(struct ListNode* head){
	struct ListNode* retl = NULL;
    
	while(head){
		struct ListNode* tmp = head;
        head = head->next;
		tmp->next = retl;
		retl = tmp;
	}
	return retl;
}
// Recursively
struct ListNode* reverseList(struct ListNode* head){
	if(NULL == head || NULL == head->next) return head;

	struct ListNode *p = head->next;
	head->next = NULL;
	struct ListNode *newhead = reverseList(p);
	p->next = head;

	return newhead;
}

File: LinkedList/Python3/E206-reverse-linked-list.py
# time:  O(n)
# space: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # My solution, recursively
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.reverse(head)
        
    def reverse(self, fhalf, shalf=None):
        if fhalf is None:
            return shalf
        f = fhalf.next
        fhalf.next = shalf
        return self.reverse(f, fhalf)

    # Iterative, faster
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = None
        while head:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr
        return prev
}
